server/src/main/java/org/opensearch/index/ShardIndexingPressure.java

```diff
@@ -0,0 +1,342 @@
+        };
+    }
+
+    public Releasable markReplicaOperationStarted(ShardId shardId, long bytes, boolean forceExecution) {
+        if(0 == bytes) { return () -> {}; }
+
+        long requestStartTime = System.nanoTime();
+        ShardIndexingPressureTracker tracker = getShardIndexingPressureTracker(shardId);
+        long nodeReplicaBytes = currentReplicaBytes.addAndGet(bytes);
+        long shardReplicaBytes = tracker.getReplicaOperationTracker().getStatsTracker().incrementCurrentBytes(bytes);
+
+        boolean shardLevelLimitBreached = false;
+        if (forceExecution == false) {
+            boolean nodeLevelLimitBreached = memoryManager.isReplicaNodeLimitBreached(tracker, nodeReplicaBytes);
+            if (nodeLevelLimitBreached == false) {
+                shardLevelLimitBreached = memoryManager.isReplicaShardLimitBreached(tracker, nodeReplicaBytes, requestStartTime);
+            }
+
+            if (shouldRejectRequest(nodeLevelLimitBreached, shardLevelLimitBreached)) {
+                replicaRejections.getAndIncrement();
+                currentReplicaBytes.addAndGet(-bytes);
+                tracker.getReplicaOperationTracker().getStatsTracker().incrementCurrentBytes(-bytes);
+                rejectShardRequest(tracker, bytes, nodeReplicaBytes, shardReplicaBytes,
+                    tracker.getReplicaOperationTracker().getRejectionTracker(), "replica");
+            }
+        }
+        totalReplicaBytes.addAndGet(bytes);
+
+        StatsTracker statsTracker = tracker.getReplicaOperationTracker().getStatsTracker();
+        markShardOperationStarted(statsTracker, tracker.getReplicaOperationTracker().getPerformanceTracker());
+        boolean isShadowModeBreach = shardLevelLimitBreached;
+
+        return () -> {
+            currentReplicaBytes.addAndGet(-bytes);
+            markShardOperationComplete(bytes, requestStartTime, isShadowModeBreach, tracker.getReplicaOperationTracker());
+            memoryManager.decreaseShardReplicaLimits(tracker);
+            tryReleaseTracker(tracker);
+        };
+    }
+
+    private boolean shouldRejectRequest(boolean nodeLevelLimitBreached, boolean shardLevelLimitBreached) {
+        return nodeLevelLimitBreached ||
+            (shardLevelLimitBreached && shardIndexingPressureSettings.isShardIndexingPressureEnforced());
+    }
+
+    private void markShardOperationStarted(StatsTracker statsTracker, PerformanceTracker performanceTracker) {
+        statsTracker.incrementRequestCount();
+        performanceTracker.incrementTotalOutstandingRequests();
+    }
+
+    private void adjustPerformanceUponCompletion(long bytes, long requestStartTime, StatsTracker statsTracker,
+                                                 PerformanceTracker performanceTracker) {
+        long requestEndTime = System.nanoTime();
+        long requestLatency = TimeUnit.NANOSECONDS.toMillis(requestEndTime - requestStartTime);
+
+        performanceTracker.addLatencyInMillis(requestLatency);
```

Comment:
Sorry why do we need to perform this computation

Reply:
We are keeping track of request latency, moving window throughputs, and total historical latency (sum) in ms across, as it provides more buffer with long data types and are more meaningful for operation level stats.

Reply:
Generally values are kept in nanos and converted back into other units lazily. 
```
IndexingStats.Stats stats(boolean isThrottled, long currentThrottleMillis) {
            return new IndexingStats.Stats(
                indexMetric.count(), TimeUnit.NANOSECONDS.toMillis(indexMetric.sum()), indexCurrent.count(), indexFailed.count(),
                deleteMetric.count(), TimeUnit.NANOSECONDS.toMillis(deleteMetric.sum()), deleteCurrent.count(),
                noopUpdates.count(), isThrottled, TimeUnit.MILLISECONDS.toMillis(currentThrottleMillis));
        }
```

Reply:
Agree for stats, however here we are also using the latency values for performing bit wise operations *(floating-point values)* during moving average calculation, and also to keep a cumulative sum of latencies over time in a long data types with delays the overflow.